feat: retro template search filter

nickoferrall opened the pull request:
Fix https://github.com/ParabolInc/parabol/issues/6791

This PR implements a search filter so users can quickly find the template they're looking for. The UI does not sparkle on mobile as there are existing template UI bugs as outlined in: https://github.com/ParabolInc/parabol/issues/5978

<details>

![template-filter](https://user-images.githubusercontent.com/39854876/174806795-610bb5ee-7425-4af1-9b50-4995692eae70.png)

</details>


### To test

- [ ] Typing in the search bar filters templates in Team, Organization, and Public
- [ ] If you begin typing in one tab, e.g. Public, and switch to the Team tab, the search query is removed

nickoferrall commented:
Hey @ackernaut, would you mind reviewing this PR and seeing if the design I went with looks good to you?

ackernaut reviewed (changes requested):
This is cool @nickoferrall !

- I see a blue outline on the search box at all times. I would style it to change from transparent to blue on focus, as that’s a focus pattern we’re using in other places (tapping on a standup response, tapping on the agenda item input, etc.)
- I noticed that the search box is still available when the team or org list is empty. Should we disable it in that case (visually muted, no input cursor)? I noticed that when the list is not empty but there aren’t any matching results, you dynamically update the help text ‘_No public templates match your search query "ninja"_‘. Should we use that pattern even for an empty list? Or should we hide the search altogether if the list includes 5 templates or fewer?
- I like that the search clears when switching tabs. Should we add a little X icon on the right that can clear it for the current tab?
- I noticed this is not on the Sprint Poker template selector which is okay, there are only 2 public templates. Thinking about this, is this one reason to hide search if there are 5 templates or fewer?

ackernaut commented:
@nickoferrall 
- Can we inset the search control 8px, use the pill radius (like Poker Scope search) or 4px radius, and then go from Slate 400 to Sky 500?
![Screen Shot 2022-06-21 at 10 26 14](https://user-images.githubusercontent.com/307286/174838729-e49fcdc0-7497-4629-bf70-cb7e934328f6.png)
![Screen Shot 2022-06-21 at 10 27 47](https://user-images.githubusercontent.com/307286/174838732-a6e72bdd-2ebc-45e6-8111-1c16e9b1b886.png)
- Hiding the search sounds complex. I think we can leave it as is for now. If the list is empty the results are going to be empty. However, I’m not sure if it’s apparent that the search is scoped to the tab. Maybe we say ‘Search [tab name] templates…’ in the placeholder text? Is that possible?
- Thanks for adding the little X if not too much creep in scope

nickoferrall commented:
@ackernaut ooh yeah, the poker scope search bar style looks much nicer here. New screenshot below:

<details>

![Screenshot 2022-06-21 at 19 28 24](https://user-images.githubusercontent.com/39854876/174872400-1e6aa787-c949-4cb7-ae66-9f5212b5d754.png)

</details>

I implemented the colour change onFocus from Slate 400 to Sky 500 but I actually think it looks better when it remains Slate 400. That's how the GitHub search bar works, so I removed the onFocus logic. Curious what you think - gif below:

<details>

![change-border-colour](https://user-images.githubusercontent.com/39854876/174872986-188fcc5a-8ca2-43a7-a8af-42d40626eb03.gif)

</details>

> Maybe we say ‘Search [tab name] templates…’ in the placeholder text?

I love this idea! Update pushed which includes the template type in the placeholder. I've also added the close search icon and only display it once there's text to clear.

ackernaut commented:
@nickoferrall wow this is really coming along!
- I think the divider bar between the icon and the search input text is unnecessary in this case. In Sprint Poker the search icon is sometimes a dropdown button so the bar probably makes it feel like a split control
- I’m okay dropping the blue focus styles in this case, since the control is conventual (outlined text box with a search icon)

mattkrick commented on packages/client/modules/meeting/components/ReflectTemplateSearchBar.tsx:
```diff
@@ -0,0 +1,139 @@
+            value={templateSearchQuery ?? ''}
+          />
+        </InputWrapper>
+        <ClearSearchIcon isEmpty={!templateSearchQuery} onClick={clearSearch}>
```
-1 the input field should still have focus after clearing

mattkrick commented on packages/client/modules/meeting/components/ReflectTemplateListTeam.tsx:
```diff
@@ -38,13 +38,30 @@ interface Props {
+  const {teamTemplates, templateSearchQuery} = settings
   const edges = teamTemplates.map((t) => ({node: {id: t.id}})) as readonly {node: {id: string}}[]
   useActiveTopTemplate(edges, activeTemplateId, teamId, isActive, 'retrospective')
+  const filteredTemplates = teamTemplates.filter(({name}) =>
```
+1 curious, could you reuse the following hook? https://github.com/ParabolInc/parabol/blob/c3a29b0323049581c88e4a3dc5313b94ca8876da/packages/client/hooks/useFilteredItems.ts

mattkrick commented on packages/client/modules/meeting/components/ReflectTemplateListTeam.tsx:
```diff
@@ -53,9 +70,12 @@ const ReflectTemplateListTeam = (props: Props) => {
   return (
     <TemplateList>
-      {teamTemplates.map((template) => {
+      {filteredTemplates.map((template) => {
```
+1 it's nice to see the search query highlighted or made bold in the result. something like this: https://github.com/ParabolInc/parabol/blob/8c68bf771a350b31c7948e6bd11c5c110e30c977/packages/client/components/TypeAheadLabel.tsx

nickoferrall replied:
Yeah, I do like seeing the text become bold when it matches the query. As the title is already bold here, I've tried highlighting it instead. I kinda like it, not sure if it feels too busy. What do you think?

![template-search](https://user-images.githubusercontent.com/39854876/175565410-79196244-ed15-4940-a3b2-19563de5d6f8.gif)

mattkrick replied:
i like it! I was gonna suggest going a shade lighter, but it passes the contrast ratio test, so I'm happy!
<img width="333" alt="Screen Shot 2022-06-24 at 11 50 51 AM" src="https://user-images.githubusercontent.com/5514175/175647250-eda12891-2205-4ac4-98d4-6db42208c047.png">
